consensus/hotstuff/timeoutaggregator/timeout_aggregator.go

```diff
@@ -0,0 +1,231 @@
+		queuedTimeoutsNotifier:     engine.NewNotifier(),
+		enteringViewEventsNotifier: engine.NewNotifier(),
+		queuedTimeouts:             queuedTimeouts,
+	}
+
+	componentBuilder := component.NewComponentManagerBuilder()
+	for i := 0; i < defaultTimeoutAggregatorWorkers; i++ { // manager for worker routines that process inbound events
+		componentBuilder.AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+			ready()
+			aggregator.queuedTimeoutsProcessingLoop(ctx)
+		})
+	}
+	componentBuilder.AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+		ready()
+		aggregator.enteringViewProcessingLoop(ctx)
+	})
+
+	aggregator.ComponentManager = componentBuilder.Build()
+	return aggregator, nil
+}
+
+// queuedTimeoutsProcessingLoop is the event loop which waits for notification about pending work
+// and as soon as there is some it triggers processing.
+// All errors are propagated to SignalerContext
+func (t *TimeoutAggregator) queuedTimeoutsProcessingLoop(ctx irrecoverable.SignalerContext) {
+	notifier := t.queuedTimeoutsNotifier.Channel()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-notifier:
+			err := t.processQueuedTimeoutEvents(ctx)
+			if err != nil {
+				ctx.Throw(fmt.Errorf("internal error processing queued timeout events: %w", err))
+				return
+			}
+		}
+	}
+}
+
+// processQueuedTimeoutEvents is a function which dispatches previously queued timeouts on worker thread
+// This function is called whenever we have queued timeouts ready to be dispatched.
+// No errors are expected during normal operations.
+func (t *TimeoutAggregator) processQueuedTimeoutEvents(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		msg, ok := t.queuedTimeouts.Pop()
+		if ok {
+			timeoutObject := msg.(*model.TimeoutObject)
+			err := t.processQueuedTimeout(timeoutObject)
+			if err != nil {
+				return fmt.Errorf("could not process pending TO %v: %w", timeoutObject.ID(), err)
+			}
+
+			t.log.Info().
+				Uint64("view", timeoutObject.View).
+				Str("timeout_id", timeoutObject.ID().String()).
+				Msg("TO has been processed successfully")
+
+			continue
+		}
+
+		// when there is no more messages in the queue, back to the loop to wait
+		// for the next incoming message to arrive.
+		return nil
+	}
+}
+
+// processQueuedTimeout performs actual processing of queued timeouts, this method is called from multiple
+// concurrent goroutines.
+func (t *TimeoutAggregator) processQueuedTimeout(timeoutObject *model.TimeoutObject) error {
+	// TODO: replace this check by a specific function to query if there is epoch by view
+	_, err := t.committee.LeaderForView(timeoutObject.View)
+	if err != nil {
+		// ignore TO if we don't have information for epoch
+		if errors.Is(err, model.ErrViewForUnknownEpoch) {
+			return nil
+		}
+		return fmt.Errorf("unknown error when querying epoch state for view %d: %w", timeoutObject.View, err)
+	}
```

Comment:
In my mind, it would be more intuitive to have this logic for handling unknown future views to be included in `TimeoutCollectors`. For `TimeoutCollectors` we should anyway specify how it would handle a call to `GetOrCreateCollector` with a view for a future unknown epoch.

Concrete suggestion:
* https://github.com/onflow/flow-go/blob/f40669d422f7f67b193c0dc8a35e95220559fd56/consensus/hotstuff/timeoutaggregator/timeout_collectors.go#L13-L14 could return an `model.ErrViewForUnknownEpoch`
* include handling of `ErrViewForUnknownEpoch` in [line 154](https://github.com/onflow/flow-go/blob/f40669d422f7f67b193c0dc8a35e95220559fd56/consensus/hotstuff/timeoutaggregator/timeout_aggregator.go#L154)

Reply:
I have moved check for epoch into `TimeoutCollectors` and now `GetOrCreateCollector` checks for both pruning height and unknown epoch error conditions. I am not really following why we need `model.ErrViewForUnknownEpoch` for factory as well?

Reply:
Sorry, I skipped this detail in my comment. Here is my reasoning why the factory method should be allowed to return a `model.ErrViewForUnknownEpoch`:
* During a factory method call, we need to query the  [`Replicas` API](https://github.com/onflow/flow-go/blob/35d88f2fc769aa3d99586fba43d4830c6cf0c132/consensus/hotstuff/committee.go#L31) to get the set of authorized consensus nodes, their keys, quorum thresholds. This all is needed to initialize the subcomponents like `TimeoutProcessor`, `TimeoutSignatureAggregator`, ...
* Obviously that is going to error if we feed a view for an unknown future epoch into the `NewCollectorFactoryMethod`. And the error source will be the `Replicas` interface. 
* From the perspective of the `NewCollectorFactoryMethod` and all the layers below, this is a completely benign error. The `ErrViewForUnknownEpoch` just says that we have asked about something `Replicas` didn't know about. 

I am saying that the code already nicely bubbles up the consistent error. We only have to document it properly. Then we don't need all these pre-checks just to make sure that the Epoch information for the view is available. For me that just seems cleaner 🤷 .



https://github.com/onflow/flow-go/blob/35d88f2fc769aa3d99586fba43d4830c6cf0c132/consensus/hotstuff/timeoutcollector/timeout_processor.go#L95-L96

Reply:
Yeah, that works, I was thinking about a separate call to committee to make it explicit but that works as well.